File: charge_control.py
"""
充电控制 - 核心充电控制模块
通过内核 sysfs 接口管理 Android 电池充电。
"""

import os
import json
import time
import logging
from datetime import datetime
from pathlib import Path

logger = logging.getLogger(__name__)

# 电池/充电控制的 sysfs 路径
SYSFS_PATHS = {
    "battery_capacity": [
        "/sys/class/power_supply/battery/capacity",
        "/sys/class/power_supply/BAT0/capacity",
    ],
    "battery_status": [
        "/sys/class/power_supply/battery/status",
        "/sys/class/power_supply/BAT0/status",
    ],
    "battery_temp": [
        "/sys/class/power_supply/battery/temp",
        "/sys/class/power_supply/BAT0/temp",
    ],
    "battery_voltage": [
        "/sys/class/power_supply/battery/voltage_now",
        "/sys/class/power_supply/BAT0/voltage_now",
    ],
    "battery_current": [
        "/sys/class/power_supply/battery/current_now",
        "/sys/class/power_supply/BAT0/current_now",
    ],
    "battery_health": [
        "/sys/class/power_supply/battery/health",
        "/sys/class/power_supply/BAT0/health",
    ],
    "charging_enabled": [
        "/sys/class/power_supply/battery/charging_enabled",
        "/sys/kernel/debug/charger/charging_enable",
        "/proc/driver/mmi_battery/charging",
    ],
    "charge_type": [
        "/sys/class/power_supply/battery/charge_type",
        "/sys/class/power_supply/usb/charge_type",
    ],
    "input_current_limit": [
        "/sys/class/power_supply/battery/input_current_limit",
        "/sys/class/power_supply/usb/input_current_limit",
    ],
    "constant_charge_current": [
        "/sys/class/power_supply/battery/constant_charge_current",
        "/sys/class/power_supply/battery/constant_charge_current_max",
    ],
    "charge_control_limit": [
        "/sys/class/power_supply/battery/charge_control_limit",
        "/sys/devices/platform/soc/soc:qti_battery_charger/charge_limit",
    ],
}

CONFIG_PATH = os.path.join(os.path.dirname(__file__), "config.json")

MODES = {
    "normal": {"max_current_ma": 2000, "max_voltage_mv": 4350},
    "fast": {"max_current_ma": 4000, "max_voltage_mv": 4400},
    "trickle": {"max_current_ma": 500, "max_voltage_mv": 4200},
    "power_saving": {"max_current_ma": 1000, "max_voltage_mv": 4300},
    "super_saver": {"max_current_ma": 300, "max_voltage_mv": 4100},
}


def _read_sysfs(key: str) -> str | None:
    """依次尝试该 sysfs 键的所有候选路径，返回第一个可读取的值。"""
    for path in SYSFS_PATHS.get(key, []):
        try:
            with open(path, "r") as f:
                return f.read().strip()
        except (OSError, PermissionError):
            continue
    return None


def _write_sysfs(key: str, value: str) -> bool:
    """将值写入指定键的第一个可写 sysfs 路径。"""
    for path in SYSFS_PATHS.get(key, []):
        try:
            with open(path, "w") as f:
                f.write(str(value))
            logger.info("已写入 %s -> %s", value, path)
            return True
        except (OSError, PermissionError):
            continue
    logger.warning("无法将 '%s' 写入键 '%s' 的任何路径", value, key)
    return False


def load_config() -> dict:
    """从 config.json 加载配置。"""
    try:
        with open(CONFIG_PATH, "r") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError) as e:
        logger.error("加载配置失败: %s", e)
        return {}


def save_config(config: dict) -> bool:
    """将配置保存到 config.json。"""
    try:
        with open(CONFIG_PATH, "w") as f:
            json.dump(config, f, indent=2)
        return True
    except OSError as e:
        logger.error("保存配置失败: %s", e)
        return False


def get_battery_status() -> dict:
    """从 sysfs 读取当前电池状态。"""
    capacity_raw = _read_sysfs("battery_capacity")
    temp_raw = _read_sysfs("battery_temp")
    voltage_raw = _read_sysfs("battery_voltage")
    current_raw = _read_sysfs("battery_current")

    # Android 系统中温度以摄氏度十分之一为单位上报
    temp_c = None
    if temp_raw is not None:
        try:
            val = int(temp_raw)
            temp_c = val / 10.0 if abs(val) > 100 else float(val)
        except ValueError:
            pass

    voltage_mv = None
    if voltage_raw is not None:
        try:
            uv = int(voltage_raw)
            voltage_mv = uv / 1000 if uv > 10000 else uv
        except ValueError:
            pass

    current_ma = None
    if current_raw is not None:
        try:
            ua = int(current_raw)
            current_ma = ua / 1000 if abs(ua) > 10000 else ua
        except ValueError:
            pass

    return {
        "capacity": int(capacity_raw) if capacity_raw is not None else _mock_capacity(),
        "status": _read_sysfs("battery_status") or "未知",
        "health": _read_sysfs("battery_health") or "未知",
        "temperature": temp_c if temp_c is not None else _mock_temperature(),
        "voltage_mv": voltage_mv,
        "current_ma": current_ma,
        "charging_enabled": _is_charging_enabled(),
        "timestamp": datetime.utcnow().isoformat() + "Z",
    }


def _mock_capacity() -> int:
    """在 sysfs 不可用时（开发/测试环境）返回模拟电量值。"""
    import random
    return random.randint(40, 90)


def _mock_temperature() -> float:
    """在 sysfs 不可用时返回模拟温度值。"""
    import random
    return round(25.0 + random.uniform(0, 15), 1)


def _is_charging_enabled() -> bool:
    val = _read_sysfs("charging_enabled")
    if val is None:
        return True  # 未知时默认视为已启用
    return val.strip() not in ("0", "false", "disabled")


def set_charging_enabled(enabled: bool) -> bool:
    """启用或禁用充电。"""
    value = "1" if enabled else "0"
    return _write_sysfs("charging_enabled", value)


def set_charge_limit(limit_percent: int) -> bool:
    """设置最大充电上限（0-100%）。"""
    if not 0 <= limit_percent <= 100:
        raise ValueError(f"充电上限必须在 0-100 之间，收到: {limit_percent}")
    config = load_config()
    config.setdefault("charging", {})["max_limit"] = limit_percent
    save_config(config)
    return _write_sysfs("charge_control_limit", str(limit_percent))


def set_charging_mode(mode: str) -> bool:
    """应用指定的充电模式（normal/fast/trickle/power_saving/super_saver）。"""
    config = load_config()
    mode_cfg = config.get("modes", {}).get(mode) or MODES.get(mode)
    if mode_cfg is None:
        raise ValueError(f"未知的充电模式: {mode}")

    current_ma = mode_cfg.get("max_current_ma")
    ok = True
    if current_ma:
        ok = ok and _write_sysfs("input_current_limit", str(current_ma * 1000))
        ok = ok and _write_sysfs("constant_charge_current", str(current_ma * 1000))

    config.setdefault("charging", {})["mode"] = mode
    save_config(config)
    return ok


def check_temperature_protection() -> dict:
    """检查温度，若过热则降低或停止充电。"""
    config = load_config()
    threshold = config.get("charging", {}).get("temperature_threshold", 40)
    critical = config.get("charging", {}).get("temperature_critical", 45)

    status = get_battery_status()
    temp = status.get("temperature", 0)

    action = "none"
    if temp >= critical:
        set_charging_enabled(False)
        action = "charging_stopped"
    elif temp >= threshold:
        set_charging_mode("trickle")
        action = "throttled_to_trickle"
    elif not _is_charging_enabled():
        # 温度下降后恢复充电
        set_charging_enabled(True)
        action = "charging_resumed"

    return {
        "temperature": temp,
        "threshold": threshold,
        "critical": critical,
        "action": action,
    }


def get_all_settings() -> dict:
    """返回当前配置与实时电池数据的合并结果。"""
    config = load_config()
    battery = get_battery_status()
    return {
        "config": config,
        "battery": battery,
    }

File: src/charge_control.c
#include "charge_control.h"
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>
#include <pthread.h>

/* ------------------------------------------------------------------ */
/* sysfs path tables (mirrors charge_control.py)                       */
/* ------------------------------------------------------------------ */

static const char *BATTERY_CAPACITY_PATHS[] = {
    "/sys/class/power_supply/battery/capacity",
    "/sys/class/power_supply/BAT0/capacity",
    NULL
};

static const char *BATTERY_STATUS_PATHS[] = {
    "/sys/class/power_supply/battery/status",
    "/sys/class/power_supply/BAT0/status",
    NULL
};

static const char *BATTERY_TEMP_PATHS[] = {
    "/sys/class/power_supply/battery/temp",
    "/sys/class/power_supply/BAT0/temp",
    NULL
};

static const char *BATTERY_VOLTAGE_PATHS[] = {
    "/sys/class/power_supply/battery/voltage_now",
    "/sys/class/power_supply/BAT0/voltage_now",
    NULL
};

static const char *BATTERY_CURRENT_PATHS[] = {
    "/sys/class/power_supply/battery/current_now",
    "/sys/class/power_supply/BAT0/current_now",
    NULL
};

static const char *BATTERY_HEALTH_PATHS[] = {
    "/sys/class/power_supply/battery/health",
    "/sys/class/power_supply/BAT0/health",
    NULL
};

static const char *CHARGING_ENABLED_PATHS[] = {
    "/sys/class/power_supply/battery/charging_enabled",
    "/sys/kernel/debug/charger/charging_enable",
    "/proc/driver/mmi_battery/charging",
    NULL
};

static const char *INPUT_CURRENT_LIMIT_PATHS[] = {
    "/sys/class/power_supply/battery/input_current_limit",
    "/sys/class/power_supply/usb/input_current_limit",
    NULL
};

static const char *CONSTANT_CHARGE_CURRENT_PATHS[] = {
    "/sys/class/power_supply/battery/constant_charge_current",
    "/sys/class/power_supply/battery/constant_charge_current_max",
    NULL
};

static const char *CHARGE_CONTROL_LIMIT_PATHS[] = {
    "/sys/class/power_supply/battery/charge_control_limit",
    "/sys/devices/platform/soc/soc:qti_battery_charger/charge_limit",
    NULL
};

/* Built-in mode defaults (matches charge_control.py) */
typedef struct {
    const char *name;
    int max_current_ma;
    int max_voltage_mv;
} mode_cfg_t;

static const mode_cfg_t BUILTIN_MODES[] = {
    { "normal",       2000, 4350 },
    { "fast",         4000, 4400 },
    { "trickle",       500, 4200 },
    { "power_saving", 1000, 4300 },
    { "super_saver",   300, 4100 },
    { NULL, 0, 0 }
};

/* Mutex for sysfs access */
static pthread_mutex_t sysfs_mutex = PTHREAD_MUTEX_INITIALIZER;

/* ------------------------------------------------------------------ */
/* Internal helpers                                                    */
/* ------------------------------------------------------------------ */

/* Read first readable path from list; buf must be at least buf_len bytes.
 * Returns 0 on success (buf filled with trimmed value), -1 if no path readable. */
static int read_sysfs(const char **paths, char *buf, size_t buf_len)
{
    FILE *f = NULL;
    size_t n = 0;
    int i = 0;

    for (i = 0; paths[i] != NULL; i++) {
        f = fopen(paths[i], "r");
        if (!f) continue;
        n = fread(buf, 1, buf_len - 1, f);
        fclose(f);
        buf[n] = '\0';
        /* strip trailing whitespace */
        while (n > 0 && (buf[n-1] == '\n' || buf[n-1] == '\r' || buf[n-1] == ' ')) {
            buf[--n] = '\0';
        }
        return 0;
    }
    return -1;
}

/* Write value to first writable path. Returns 0 on success, -1 on failure. */
static int write_sysfs(const char **paths, const char *value)
{
    FILE *f = NULL;
    int i = 0;

    for (i = 0; paths[i] != NULL; i++) {
        f = fopen(paths[i], "w");
        if (!f) continue;
        if (fputs(value, f) != EOF) {
            fclose(f);
            return 0;
        }
        fclose(f);
    }
    return -1;
}

static int is_charging_enabled(void)
{
    char buf[32];
    if (read_sysfs(CHARGING_ENABLED_PATHS, buf, sizeof(buf)) != 0) {
        return 1; /* unknown: assume enabled */
    }
    if (strcmp(buf, "0") == 0 || strcmp(buf, "false") == 0 || strcmp(buf, "disabled") == 0) {
        return 0;
    }
    return 1;
}

/* ------------------------------------------------------------------ */
/* Public API                                                          */
/* ------------------------------------------------------------------ */

int cc_get_battery_status(battery_status_t *out)
{
    char buf[64];
    time_t now;
    struct tm tm_utc;

    if (!out) return -1;

    memset(out, 0, sizeof(*out));
    out->capacity    = -1;
    out->temperature = -1.0f;
    out->voltage_mv  = -1.0f;
    out->current_ma  = -1.0f;
    strncpy(out->status, "Unknown", sizeof(out->status) - 1);
    strncpy(out->health, "Unknown", sizeof(out->health) - 1);

    pthread_mutex_lock(&sysfs_mutex);

    /* capacity */
    if (read_sysfs(BATTERY_CAPACITY_PATHS, buf, sizeof(buf)) == 0) {
        out->capacity = atoi(buf);
    }
    /* Bug fix: sysfs unavailable → return -1, never a random value */

    /* status */
    if (read_sysfs(BATTERY_STATUS_PATHS, buf, sizeof(buf)) == 0) {
        snprintf(out->status, sizeof(out->status), "%s", buf);
    }

    /* health */
    if (read_sysfs(BATTERY_HEALTH_PATHS, buf, sizeof(buf)) == 0) {
        snprintf(out->health, sizeof(out->health), "%s", buf);
    }

    /* temperature: Android reports in tenths of degree Celsius */
    if (read_sysfs(BATTERY_TEMP_PATHS, buf, sizeof(buf)) == 0) {
        long val = atol(buf);
        if (val > 100 || val < -100) {
            out->temperature = (float)val / 10.0f;
        } else {
            out->temperature = (float)val;
        }
    }

    /* voltage: microvolts → millivolts */
    if (read_sysfs(BATTERY_VOLTAGE_PATHS, buf, sizeof(buf)) == 0) {
        long uv = atol(buf);
        if (uv > 10000) {
            out->voltage_mv = (float)uv / 1000.0f;
        } else {
            out->voltage_mv = (float)uv;
        }
    }

    /* current: microamperes → milliamperes */
    if (read_sysfs(BATTERY_CURRENT_PATHS, buf, sizeof(buf)) == 0) {
        long ua = atol(buf);
        if (ua > 10000 || ua < -10000) {
            out->current_ma = (float)ua / 1000.0f;
        } else {
            out->current_ma = (float)ua;
        }
    }

    out->charging_enabled = is_charging_enabled();

    pthread_mutex_unlock(&sysfs_mutex);

    /* timestamp (UTC ISO-8601) */
    time(&now);
    gmtime_r(&now, &tm_utc);
    strftime(out->timestamp, sizeof(out->timestamp), "%Y-%m-%dT%H:%M:%SZ", &tm_utc);

    return 0;
}

int cc_set_charging_enabled(int enabled)
{
    int ret;
    const char *value = enabled ? "1" : "0";
    pthread_mutex_lock(&sysfs_mutex);
    ret = write_sysfs(CHARGING_ENABLED_PATHS, value);
    pthread_mutex_unlock(&sysfs_mutex);
    return ret;
}

int cc_set_charge_limit(int limit_percent)
{
    char buf[16];
    int ret;

    if (limit_percent < 0 || limit_percent > 100) return -1;

    snprintf(buf, sizeof(buf), "%d", limit_percent);
    pthread_mutex_lock(&sysfs_mutex);
    ret = write_sysfs(CHARGE_CONTROL_LIMIT_PATHS, buf);
    pthread_mutex_unlock(&sysfs_mutex);
    return ret;
}

int cc_set_charging_mode(const char *mode, cJSON *cfg)
{
    int max_current_ma = 0;
    int i;
    char buf[32];
    int ok1 = 0, ok2 = 0;

    if (!mode) return -1;

    /* Look up in config first, then fall back to builtin table */
    if (cfg) {
        cJSON *modes_obj = cJSON_GetObjectItem(cfg, "modes");
        if (modes_obj) {
            cJSON *mode_obj = cJSON_GetObjectItem(modes_obj, mode);
            if (mode_obj) {
                cJSON *cur = cJSON_GetObjectItem(mode_obj, "max_current_ma");
                if (cJSON_IsNumber(cur)) max_current_ma = cur->valueint;
            }
        }
    }

    if (max_current_ma == 0) {
        for (i = 0; BUILTIN_MODES[i].name != NULL; i++) {
            if (strcmp(BUILTIN_MODES[i].name, mode) == 0) {
                max_current_ma = BUILTIN_MODES[i].max_current_ma;
                break;
            }
        }
    }

    if (max_current_ma == 0) return -1; /* unknown mode */

    snprintf(buf, sizeof(buf), "%d", max_current_ma * 1000);

    pthread_mutex_lock(&sysfs_mutex);
    ok1 = write_sysfs(INPUT_CURRENT_LIMIT_PATHS, buf);
    ok2 = write_sysfs(CONSTANT_CHARGE_CURRENT_PATHS, buf);
    pthread_mutex_unlock(&sysfs_mutex);

    /* Return 0 if at least one write succeeded (mirrors Python behaviour) */
    return (ok1 == 0 || ok2 == 0) ? 0 : -1;
}

int cc_check_temperature_protection(cJSON *cfg, float *out_temp, const char **out_action)
{
    battery_status_t bat;
    float threshold = 40.0f;
    float critical  = 45.0f;
    static const char *action = "none";

    if (cfg) {
        cJSON *charging = cJSON_GetObjectItem(cfg, "charging");
        if (charging) {
            cJSON *thr = cJSON_GetObjectItem(charging, "temperature_threshold");
            cJSON *cri = cJSON_GetObjectItem(charging, "temperature_critical");
            if (cJSON_IsNumber(thr)) threshold = (float)thr->valuedouble;
            if (cJSON_IsNumber(cri)) critical  = (float)cri->valuedouble;
        }
    }

    if (cc_get_battery_status(&bat) != 0) return -1;

    if (out_temp) *out_temp = bat.temperature;

    if (bat.temperature >= critical) {
        cc_set_charging_enabled(0);
        action = "charging_stopped";
    } else if (bat.temperature >= threshold) {
        cc_set_charging_mode("trickle", cfg);
        action = "throttled_to_trickle";
    } else if (!bat.charging_enabled) {
        cc_set_charging_enabled(1);
        action = "charging_resumed";
    } else {
        action = "none";
    }

    if (out_action) *out_action = action;
    return 0;
}
